File: src/vmm.c
/*
 * Copyright 2019, Data61, CSIRO (ABN 41 687 119 230)
 * Copyright 2022, UNSW (ABN 57 195 873 179)
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */
#include <stddef.h>
#include <sel4cp.h>
#include "util/util.h"
#include "vgic/vgic.h"
#include "vmm.h"
#include "linux.h"
#include "tcb.h"
#include "vcpu.h"
#include "fault.h"
#include "guest.h"
#include "virq.h"

/* Data for the guest's kernel image. */
extern char _guest_kernel_image[];
extern char _guest_kernel_image_end[];
/* Data for the device tree to be passed to the kernel. */
extern char _guest_dtb_image[];
extern char _guest_dtb_image_end[];
/* Data for the initial RAM disk to be passed to the kernel. */
extern char _guest_initrd_image[];
extern char _guest_initrd_image_end[];
/* seL4CP will set this variable to the start of the guest RAM memory region. */
uintptr_t guest_ram_vaddr;

static void serial_ack(size_t vcpu_id, int irq, void *cookie) {
    /*
     * For now we by default simply ack the serial IRQ, we have not
     * come across a case yet where more than this needs to be done.
     */
    sel4cp_irq_ack(SERIAL_IRQ_CH);
}

void
init(void)
{
    // Initialise the VMM, the VCPU(s), and start the guest
    LOG_VMM("starting \"%s\"\n", sel4cp_name);
    // Place all the binaries in the right locations before starting the guest
    size_t kernel_size = _guest_kernel_image_end - _guest_kernel_image;
    size_t dtb_size = _guest_dtb_image_end - _guest_dtb_image;
    size_t initrd_size = _guest_initrd_image_end - _guest_initrd_image;
    uintptr_t kernel_pc = linux_setup_images(guest_ram_vaddr,
                                      (uintptr_t) _guest_kernel_image,
                                      kernel_size,
                                      (uintptr_t) _guest_dtb_image,
                                      GUEST_DTB_VADDR,
                                      dtb_size,
                                      (uintptr_t) _guest_initrd_image,
                                      GUEST_INIT_RAM_DISK_VADDR,
                                      initrd_size
                                      );
    if (!kernel_pc) {
        LOG_VMM_ERR("Failed to initialise guest images\n");
        return;
    }
    // Initialise the virtual GIC driver
    bool success = virq_controller_init(GUEST_VCPU_ID);
    if (!success) {
        LOG_VMM_ERR("Failed to initialise emulated interrupt controller\n");
        return;
    }
    // @ivanv: Note that remove this line causes the VMM to fault if we
    // actually get the interrupt. This should be avoided by making the VGIC driver more stable.
    success = virq_register(GUEST_VCPU_ID, SERIAL_IRQ, &serial_ack, NULL);
    // Just in case there is already an interrupt available to handle, we ack it here.
    // @ivanv: not sure if this is necessary? is this the right place to do this
    sel4cp_irq_ack(SERIAL_IRQ_CH);
    /* Finally start the guest */
    guest_start(GUEST_VCPU_ID, kernel_pc, GUEST_DTB_VADDR, GUEST_INIT_RAM_DISK_VADDR);
}

void
notified(sel4cp_channel ch)
{
    switch (ch) {
        case SERIAL_IRQ_CH: {
            bool success = virq_inject(GUEST_VCPU_ID, SERIAL_IRQ);
            if (!success) {
                LOG_VMM_ERR("IRQ %d dropped on vCPU %d\n", SERIAL_IRQ, GUEST_VCPU_ID);
            }
            break;
        }
        default:
            printf("Unexpected channel, ch: 0x%lx\n", ch);
    }
}

void
fault(sel4cp_id id, sel4cp_msginfo msginfo)
{
    if (id != GUEST_VCPU_ID) {
        LOG_VMM_ERR("Unexpected faulting PD/VM with id %d\n", id);
        return;
    }
    // This is the primary fault handler for the guest, all faults that come
    // from seL4 regarding the guest will need to be handled here.
    uint64_t label = sel4cp_msginfo_get_label(msginfo);
    bool success = false;
    switch (label) {
        case seL4_Fault_VMFault:
            success = handle_vm_fault(id);
            break;
        case seL4_Fault_UnknownSyscall:
            success = handle_unknown_syscall(id);
            break;
        case seL4_Fault_UserException:
            success = handle_user_exception(id);
            break;
        case seL4_Fault_VGICMaintenance:
            success = handle_vgic_maintenance(id);
            break;
        case seL4_Fault_VCPUFault:
            success = handle_vcpu_fault(id);
            break;
        case seL4_Fault_VPPIEvent:
            success = handle_vppi_event(id);
            break;
        default:
            LOG_VMM_ERR("unknown fault, stopping VM with ID %d\n", id);
            sel4cp_vm_stop(id);
            return;
            // @ivanv: print out the actual fault details
    }

    if (success) {
        /* Now that we have handled the fault, we reply to it so that the guest can resume execution. */
        reply_to_fault();
    } else {
        LOG_VMM_ERR("Failed to handle %s fault\n", fault_to_string(label));
    }
}

File: examples/zig/src/vmm.zig
// Copyright 2023, UNSW
// SPDX-License-Identifier: BSD-2-Clause

const std = @import("std");
const c = @cImport({
    @cInclude("microkit.h");
    @cInclude("libvmm/virq.h");
    @cInclude("libvmm/guest.h");
    @cInclude("libvmm/arch/aarch64/linux.h");
});

// In this example we only have one virtual CPU
const GUEST_BOOT_VCPU_ID = 0;
// There are the hard-coded addresses that both the VMM and Linux guest need
// to be aware of. For example the address of the DTB and initial RAM disk are
// passed to Linux when booting it.
const GUEST_RAM_VADDR: usize = 0x40000000;
const GUEST_DTB_VADDR: usize = 0x4f000000;
const GUEST_INIT_RAM_DISK_VADDR: usize = 0x4d700000;
const GUEST_RAM_SIZE: usize = 0x10000000;

// Below we make use of Zig's '@embedFile' builtin functions to easily include
// the arficats/images that the VMM needs. You will notice these aren't based
// on a relative or absolute path, this is because the build script enables us
// to find the files just based on the name by adding each one as a 'module'.

// Data for the guest's kernel image.
const guest_kernel_image = @embedFile("linux");
// Data for the device tree to be passed to the kernel.
const guest_dtb_image = @embedFile("dtb");
// Data for the initial RAM disk to be passed to the kernel.
const guest_initrd_image = @embedFile("initrd");

// In Zig the standard library comes with printf-like functionality with the
// ability to provide your own function to ouput the characters. This is
// extremely useful for us! Without changing the standard library or bringing
// in a 3rd party library, in a dozen lines of code we can get formatted
// printing functionality that ends up calling to microkit_dbg_puts which then
// outputs to the platform's serial connection.
//
// In a production system you might want to instead hook up the logger to a
// user-level UART driver, but for the example we only need logging for debug
// mode.
const log = struct {
    var writer: std.Io.Writer = .{
        // Having an empty buffer means the writer flushes everything out
        // immediately.
        .buffer = &.{},
        .vtable = &.{
            .drain = debug_uart_put_str
        }
    };

    fn debug_uart_put_str(_: *std.Io.Writer, data: []const []const u8, splat: usize) std.Io.Writer.Error!usize {
        var written: usize = 0;
        for (data, 0..) |str, i| {
            for (str) |ch| {
                c.microkit_dbg_putc(ch);
            }
            written += str.len;

            // If we're on the last item of data, we must print it `splat`
            // times and therefore repeat it `splat - 1` times.
            if (i == data.len - 1) {
                for (0..splat - 1) |_| {
                    for (str) |ch| {
                        c.microkit_dbg_putc(ch);
                    }
                    written += str.len;
                }
            }
        }

        return written;
    }

    pub fn info(comptime fmt: []const u8, args: anytype) void {
        writer.print("VMM|INFO: " ++ fmt ++ "\n", args) catch {};
    }

    pub fn err(comptime fmt: []const u8, args: anytype) void {
        writer.print("VMM|ERROR: " ++ fmt ++ "\n", args) catch {};
    }
};

const SERIAL_IRQ_CH: c.microkit_channel = 1;
const SERIAL_IRQ: i32 = 33;

fn serial_ack(_: usize, _: c_int, _: ?*anyopaque) callconv(.c) void {
    // Nothing else needs to be done other than acking the IRQ.
    c.microkit_irq_ack(SERIAL_IRQ_CH);
}

export fn init() callconv(.c) void {
    // Initialise the VMM, the VCPU(s), and start the guest
    log.info("starting", .{});
    // Place all the binaries in the right locations before starting the guest
    const kernel_pc = c.linux_setup_images(
                GUEST_RAM_VADDR,
                @intFromPtr(guest_kernel_image),
                guest_kernel_image.len,
                @intFromPtr(guest_dtb_image),
                GUEST_DTB_VADDR,
                guest_dtb_image.len,
                @intFromPtr(guest_initrd_image),
                GUEST_INIT_RAM_DISK_VADDR,
                guest_initrd_image.len
            );
    if (kernel_pc == 0) {
        log.err("Failed to initialise guest images\n", .{});
        return;
    }
    // Initialise the virtual interrupt controller
    if (!c.virq_controller_init()) {
        log.err("Failed to initialise virtual interrupt controller\n", .{});
        return;
    }
    // Register the interrupt for the UART with the virtual interrupt controller
    if (!c.virq_register(GUEST_BOOT_VCPU_ID, SERIAL_IRQ, &serial_ack, null)) {
        log.err("Failed to register serial IRQ\n", .{});
        return;
    }
    // Just in case there is already an interrupt from the UART available to
    // handle, we ack it here.
    c.microkit_irq_ack(SERIAL_IRQ_CH);
    // Finally we can start the guest
    if (!c.guest_start(kernel_pc, GUEST_DTB_VADDR, GUEST_INIT_RAM_DISK_VADDR)) {
        log.err("Failed to start guest\n", .{});
        return;
    }
}

export fn notified(ch: c.microkit_channel) callconv(.c) void {
    switch (ch) {
        SERIAL_IRQ_CH => {
            const success = c.virq_inject(SERIAL_IRQ);
            if (!success) {
                log.err("IRQ {x} dropped\n", .{ SERIAL_IRQ });
            }
        },
        else => log.err("Unexpected channel, ch: {x}\n", .{ ch })
    }
}

extern fn fault_handle(id: c.microkit_child, msginfo: c.microkit_msginfo) callconv(.c) bool;

export fn fault(id: c.microkit_child, msginfo: c.microkit_msginfo, msginfo_reply: *c.microkit_msginfo) callconv(.c) bool {
    if (fault_handle(id, msginfo)) {
        // Now that we have handled the fault, we reply to it so that the guest can resume execution.
        msginfo_reply.* = c.microkit_msginfo_new(0, 0);
        return true;
    } else {
        log.err("Failed to handle fault\n", .{});
        return false;
    }
}
